### DISCUSSION TITLE
Has anyone built an app with teams or invite functionality?

### DISCUSSION THREAD

**Discussion by chrtze** [nhost/nhost#1100] [Q&A] [answered]
Hi everyone! We are happy users of nhost for our [subscriber platform](https://pro.reactflow.dev/). Currently, we are looking at how to build a teams/invitation functionality where registered and subscribed users can invite team members which will then be added to their team and have access to the same content.

Considerations and open questions so far:

* how to create or invite users through another user in nhost?
* how to configure the permissions for team members so that they can see content from the parent but not the billing details etc.?
* how to decide who is the admin for a team?
* how to send out invitation mails and join the team automatically when signing up?

Has anyone of the community built something similar using nhost? I would be super interested to learn more about example implementations and you ideas 🚀

**Comment by elitan** [comment, marked as answer]
Here's one suggestion.

## Tables

```
teams
- id
- name

team_members
- id
- team_id FK to team.id
- user_id FK to auth.user.id (nullable)
- type (member, administrator, owner)
- invite_email (must be of type CITEXT
- created_by_user_id FK to auth.user.id
```

## Flow

A user that is an `administrator` or `owner` of a team can invite another user by inserting their email to `team_members`.

If `team_members.user_id` is `null`, it's a team member invite. When `team_members.user_id` is not `null`, the user is a member.

## Permisson for `team_invites`

### Insert

Owners and administrators of a team are allowed to invite new team members.

```json
{
  "team": {
    "team_members": {
      "_and": [
        {
          "user_id": {
            "_eq": "X-Hasura-User-Id"
          }
        },
        {
          "type": {
            "_in": [
              "administrator",
              "owner"
            ]
          }
        }
      ]
    }
  }
}
```

Allowed to insert: 

```
- email
- type
- team_id
```

### Select

Users are allowed to see invites if one of the following is true:

- The user's email is the same as the invited user
- The user is an owner or administrator of the team

```json
{
  "_or": [
    {
      "_exists": {
        "_table": {
          "schema": "auth",
          "name": "users"
        },
        "_where": {
          "_and": [
            {
              "id": {
                "_eq": "X-Hasura-User-Id"
              }
            },
            {
              "email": {
                "_ceq": [
                  "email"
                ]
              }
            }
          ]
        }
      }
    },
    {
      "team": {
        "team_members": {
          "_and": [
            {
              "user_id": {
                "_eq": "X-Hasura-User-Id"
              }
            },
            {
              "type": {
                "_in": [
                  "owner",
                  "administrator"
                ]
              }
            }
          ]
        }
      }
    }
  ]
}
```

Allowed to select: 

- All columns.


Event trigger:

- ON INSERT to send an invite email to `invite_email`.

### Update

Users who are invited can update their team invite by setting `user_id` to their own user id. This is how a user accepts a team invite.

PRE check

```json
{
  "_exists": {
    "_table": {
      "name": "users",
      "schema": "auth"
    },
    "_where": {
      "_and": [
        {
          "id": {
            "_eq": "X-Hasura-User-Id"
          }
        },
        {
          "email": {
            "_ceq": [
              "$",
              "invite_email"
            ]
          }
        }
      ]
    }
  }
}
```

POST check:

```json
{
  "user_id": {
    "_eq": "X-Hasura-User-Id"
  }
}
```

Columns to update:

```
- user_id
```

### Delete

Owners and administrators of a team are allowed to delete team invites (and team members).

```json
{
  "team": {
    "team_members": {
      "_and": [
        {
          "user_id": {
            "_eq": "X-Hasura-User-Id"
          }
        },
        {
          "type": {
            "_in": [
              "owner",
              "administrator"
            ]
          }
        }
      ]
    }
  }
}
```


> How to configure the permissions for team members so that they can see content from the parent but not the billing details etc.?

Check the `type` for the current user in the `team_members` table.

> how to decide who is the admin for a team?

Again, you can use the `type` in the `team_members` table. This way you can have different member types. In my example above, I'm using `owner`, `administrator`, and `member`.

> how to send out invitation mails and join the team automatically when signing up?

Using an INSERT event trigger on the `team_members` table. I suggest showing a modal when users sign in showing pending team member invites that they can accept.

**Reply by chrtze** [reply]
Wow, thank you so much @elitan! I will need some time to check out all the details but this is such a valuable input for us. 👏👏👏

**Reply by st3phan** [reply]
This is more or less how I do it too.

The only difference is that I update the `user_id` in the `team_members` table, using an event trigger and function, after a user has signed up. This way you can have a  `team_members.user_id -> users.id` relationship on the `team_members` table.
